packager/media/formats/mp2t/ts_muxer.cc

```diff
@@ -21,7 +23,12 @@ Status TsMuxer::InitializeMuxer() {
   if (streams().size() > 1u)
     return Status(error::MUXER_FAILURE, "Cannot handle more than one streams.");
 
-  segmenter_.reset(new TsSegmenter(options(), muxer_listener()));
+  if (options().segment_template.empty()) {
```

Comment:
Prefer using the approach in PackedAudio: https://github.com/google/shaka-packager/tree/56908a83a76c85450d6b5834f3aa16e0660f4648/packager/media/formats/packed_audio.

Here is the summary:

- Having only one segmenter class instead of having one for SingleSegment and one for MultiSegment.
- The segmenter class writes the data into a temporary buffer first, which is already the case in the current TS segmenter implementation.
- Do not output the temporary buffer to the file in the segmenter, but in the muxer instead.
- In the muxer, the code can decide whether to write to a single file or a segment file based on the configuration. See https://github.com/google/shaka-packager/blob/56908a83a76c85450d6b5834f3aa16e0660f4648/packager/media/formats/packed_audio/packed_audio_writer.cc#L100 for an example.
- The muxer listener events are handled in the muxer. There is no need to handle that in the segmenter.

I think the code is cleaner and easier to maintain with this structure. 

What do you think?

Reply:
Hi @kqyang, thanks for the review. Yes that makes sense. Let me try it out.